openhands-sdk/openhands/sdk/security/security_service.py

```diff
@@ -0,0 +1,110 @@
+from abc import ABC, abstractmethod
+from typing import TYPE_CHECKING
+
+from openhands.sdk.utils.models import DiscriminatedUnionMixin
+
+
+if TYPE_CHECKING:
+    from openhands.sdk.conversation.state import ConversationState
+
+from openhands.sdk.event.llm_convertible.action import ActionEvent
+from openhands.sdk.security import risk
+from openhands.sdk.security.llm_analyzer import LLMSecurityAnalyzer
+from openhands.sdk.tool.builtins.finish import FinishAction
+from openhands.sdk.tool.builtins.think import ThinkAction
+
+
+class SecurityService(DiscriminatedUnionMixin, ABC):
+    """
+    Security service interface defining core security-related methods.
+    """
+
+    @abstractmethod
+    def requires_confirmation(self, action_events: list[ActionEvent]) -> bool:
+        """
+        Determine whether user confirmation is required to proceed with actions.
+
+        :param action_events: List of pending action events
+        :return: True if confirmation is needed, False otherwise
+        """
+        pass
+
+
+class SecurityServiceBase(SecurityService):
+    def __init__(
+        self,
+        state: "ConversationState",
+    ):
+        self._state = state
+
+    def requires_confirmation(
+        self,
+        action_events: list[ActionEvent],
+    ) -> bool:
+        """
+        Decide whether user confirmation is needed to proceed.
+
+        Rules:
+            1. Confirmation mode is enabled
+            2. Every action requires confirmation
+            3. A single `FinishAction` never requires confirmation
+            4. A single `ThinkAction` never requires confirmation
+        """
+        # If there are no actions there is nothing to confirm
+        if len(action_events) == 0:
+            return False
+
+        if all(
+            isinstance(action_event.action, (FinishAction, ThinkAction))
+            for action_event in action_events
+        ):
+            return False
+        # If a security analyzer is registered, use it to grab the risks of the actions
+        # involved. If not, we'll set the risks to UNKNOWN.
+        if self._state.security_analyzer is not None:
+            risks = (
+                r
+                for _, r in self._state.security_analyzer.analyze_pending_actions(
+                    action_events
+                )
+            )
+        else:
+            risks = (risk.SecurityRisk.UNKNOWN for _ in action_events)
+
+        return any(self._state.confirmation_policy.should_confirm(r) for r in risks)
+
+
+class DefaultSecurityService(SecurityServiceBase):
+    def __init__(self, state: "ConversationState"):
```

Comment:
```suggestion
    def __init__(
        self, 
        security_analyzer: SecurityAnalyzerBase | None,
        confirmation_policy: ConfirmationPolicy
    ):
```

Let's de-couple the security service as much as possible. Ideally we don't pass the entire state object. The security service should return a typed object which should include the decision, that the agent can use to modify its state properly 

```
class SecurityService(ABC):
    @abstractmethod
    def assess_actions(self, action_events: list[ActionEvent]) -> SecurityAssessment:
        """
        Assess security implications of actions.
        """
        pass
```